### PR TITLE
FEATURE: Introduce support for link_url annotation

### PR DESCRIPTION
**Pull request by davidtaylorhq**
Previously, we had support for 'grafana_path' and 'logs_url', but these were implemented in a CDCK-specific way. This commit removes support for those, and adds a more generic system. Alerts can now include `link_url` and `link_text` annotations which will be used in the alert UI.

Three new regex site settings are introduced which can be configured to enable special handling for grafana, kibana and prometheus links. This special handling will automatically add the alert timestamps to the URLs. To disable this special handling, set the regex to an empty string.

The existing 'graph_url' column is renamed to 'generator_url' to match prometheus terminology and avoid any confusion with grafana URLs.

The existing grafana_url and logs_url columns are ignored, and will be dropped in a future commit. Data has been automatically migrated to the link_url column.

### CODE REVIEW

**Comment by davidtaylorhq** [comment]
(The failing test is a CI configuration issue)

File: assets/javascripts/discourse/widgets/alert-receiver.js.es6 [hunk 1]
```diff
@@ -152,34 +152,94 @@ createWidget("alert-receiver-row", {
...
+    }
+  },
+
+  checkMatch(url, regexString) {
+    if (!regexString || !url) {
+      return false;
+    }
+    if (regexString === "" || url === "") {
+      return false;
+    }
```
Comment (CvX) [inline assets/javascripts/discourse/widgets/alert-receiver.js.es6:200]: We're not in <del>Kansas</del> <ins>ruby</ins> anymore. 😉

```suggestion
    if (!regexString || !url) {
      return false;
    }
```

File: assets/javascripts/discourse/widgets/alert-receiver.js.es6 [hunk 2]
```diff
@@ -152,34 +152,94 @@ createWidget("alert-receiver-row", {
...
-    const start = attrs.alert.starts_at;
-    const end = attrs.alert.ends_at || new Date().toISOString();
-    return `${base}#/discover?_g=(time:(from:'${start}',mode:absolute,to:'${end}'))`;
+
+    // Kibana stores its data after the # in the URL
+    // So some custom parsing is required
+    const fragment = url.hash;
+    const parts = fragment.split("?", 2);
+
+    const searchParams = new URLSearchParams(parts[1] || "");
```
Comment (CvX) [inline assets/javascripts/discourse/widgets/alert-receiver.js.es6:224]:
```suggestion
    const searchParams = new URLSearchParams(parts[1]);
```

Constructor's parameter in `URLSearchParams` is optional

File: assets/javascripts/discourse/widgets/alert-receiver.js.es6 [hunk 3]
```diff
@@ -152,34 +152,94 @@ createWidget("alert-receiver-row", {
...
+    searchParams.set(
+      "_g",
+      `(time:(from:'${starts_at}',mode:absolute,to:'${ends_at}'))`
+    );
+
+    url.hash = `${parts[0]}?${searchParams.toString()}`;
```
Comment (CvX) [inline assets/javascripts/discourse/widgets/alert-receiver.js.es6:230]:
```suggestion
    url.hash = `${parts[0]}?${searchParams}`;
```

Expressions in string templates that are not strings are converted to string by calling `toString()` on them

File: config/settings.yml [hunk 4]
```diff
@@ -6,3 +6,12 @@ plugins:
...
   prometheus_alert_receiver_debug_enabled: false
+  prometheus_alert_receiver_grafana_regex:
+    default: ''
+    client: true
+  prometheus_alert_receiver_kibana_regex:
+    default: '\/app\/kibana'
+    client: true
+  prometheus_alert_receiver_prometheus_regex:
+    default: '\/graph\?g0\.expr='
+    client: true
```
Comment (CvX) [inline config/settings.yml:17]: Needs a newline at the end (one of the linters will complain about it at some point)

**Review by CvX — CHANGES_REQUESTED** [review]
Really cool stuff! 😃 I noted just some lint-level stuff, otherwise looks good to go!

File: assets/javascripts/discourse/widgets/alert-receiver.js.es6 [hunk 5]
```diff
@@ -152,34 +152,94 @@ createWidget("alert-receiver-row", {
...
 
   transform(attrs) {
+    const starts_at = attrs.alert.starts_at;
+    const ends_at = attrs.alert.ends_at || new Date().toISOString();
+
+    let linkText = this.attrs.alert.link_text;
+    if (linkText === "") {
+      linkText = null;
+    }
+    linkText = linkText || I18n.t("prom_alert_receiver.actions.link");
```
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply assets/javascripts/discourse/widgets/alert-receiver.js.es6:162]: Right! Empty string is falsey in javascript, this is much better - thanks! 😅

File: assets/javascripts/discourse/widgets/alert-receiver.js.es6 [hunk 6]
```diff
@@ -152,34 +152,94 @@ createWidget("alert-receiver-row", {
   tagName: "tr",
 
   transform(attrs) {
+    const starts_at = attrs.alert.starts_at;
+    const ends_at = attrs.alert.ends_at || new Date().toISOString();
```
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply assets/javascripts/discourse/widgets/alert-receiver.js.es6]: 🤕 yep, fixed! The server is serializing things with snake case, and it ended up refactoring into the JS. Fixed now 👍

**Review by CvX — APPROVED** [review]

**Pull request merged**